core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -836,64 +840,86 @@ private boolean containedInAdapter(NonBlockingStore nonBlockingStore, String ada
    }
 
    private boolean allowLoad(StoreStatus storeStatus, boolean localInvocation, boolean includeStores) {
-      return !storeStatus.characteristics.contains(Characteristic.WRITE_ONLY) && (localInvocation || !isLocalOnlyLoader(storeStatus.store)) &&
-             (includeStores || storeStatus.characteristics.contains(Characteristic.READ_ONLY) || storeStatus.config.ignoreModifications());
+      return !storeStatus.hasCharacteristic(Characteristic.WRITE_ONLY) &&
+             (localInvocation || !isLocalOnlyLoader(storeStatus.store)) &&
+             (includeStores || storeStatus.hasCharacteristic(Characteristic.READ_ONLY) ||
+              storeStatus.config.ignoreModifications());
    }
 
-   private boolean isLocalOnlyLoader(NonBlockingStore store) {
+   private boolean isLocalOnlyLoader(NonBlockingStore<?, ?> store) {
       if (store instanceof LocalOnlyCacheLoader) return true;
-      NonBlockingStore unwrappedStore;
+      NonBlockingStore<?, ?> unwrappedStore;
       if (store instanceof DelegatingNonBlockingStore) {
-         unwrappedStore = ((DelegatingNonBlockingStore) store).delegate();
+         unwrappedStore = ((DelegatingNonBlockingStore<?, ?>) store).delegate();
       } else {
          unwrappedStore = store;
       }
       if (unwrappedStore instanceof LocalOnlyCacheLoader) {
          return true;
       }
       if (unwrappedStore instanceof NonBlockingStoreAdapter) {
-         return ((NonBlockingStoreAdapter) unwrappedStore).getActualStore() instanceof LocalOnlyCacheLoader;
+         return ((NonBlockingStoreAdapter<?, ?>) unwrappedStore).getActualStore() instanceof LocalOnlyCacheLoader;
       }
       return false;
    }
 
    @Override
-   public CompletionStage<Long> size(Predicate<? super StoreConfiguration> predicate) {
+   public CompletionStage<Long> approximateSize(Predicate<? super StoreConfiguration> predicate, IntSet segments) {
       long stamp = acquireReadLock();
       try {
-         checkStoreAvailability();
-         if (log.isTraceEnabled()) {
-            log.tracef("Obtaining size from stores");
+         if (!isEnabled() || !isAvailable()) {
```

Comment:
This should work outside the read lock as the variables are volatile.

Reply:
I think for availability I would have to check again after acquiring the read lock.
But I could move the `isEnabled()` check out, for the benefit of caches that will never have a store.